Sources/X509/PromiseAndFuture.swift

```diff
@@ -0,0 +1,123 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftCertificates open source project
+//
+// Copyright (c) 2023 Apple Inc. and the SwiftCertificates project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftCertificates project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+// MARK: - Promise
+final class Promise<Value, Failure: Error> {
+    private enum State {
+        case unfulfilled(observers: [CheckedContinuation<Result<Value, Failure>, Never>])
+        case fulfilled(Result<Value, Failure>)
+    }
+
+    private let state = LockedValueBox(State.unfulfilled(observers: []))
+
+    init() {}
+
+    fileprivate var result: Result<Value, Failure> {
+        get async {
+            self.state.lock()
+
+            switch self.state.unsafeUnlockedValue {
+            case .fulfilled(let result):
+                defer { self.state.unlock() }
+                return result
+
+            case .unfulfilled(var observers):
+                return await withCheckedContinuation {
+                    (continuation: CheckedContinuation<Result<Value, Failure>, Never>) in
+                    observers.append(continuation)
+                    self.state.unsafeUnlockedValue = .unfulfilled(observers: observers)
+                    self.state.unlock()
+                }
+            }
+        }
+    }
+
+    func fulfil(with result: Result<Value, Failure>) {
+        self.state.withLockedValue { state in
+            switch state {
+            case .fulfilled(let oldResult):
+                fatalError("tried to fulfil Promise that is already fulfilled to \(oldResult). New result: \(result)")
+            case .unfulfilled(let observers):
+                for observer in observers {
+                    observer.resume(returning: result)
+                }
+                state = .fulfilled(result)
+            }
+        }
+    }
+
+    deinit {
+        switch self.state.unsafeUnlockedValue {
```

Comment:
I don't think this optimization is appropriate: we can just take the lock, it'll be uncontended.